example_scripts/Building_blocks/Panoptes_client_examples/subject_downloader.py

```diff
@@ -0,0 +1,88 @@
+ subject sets, but appeared to work well.  Hopefully the OS related statements are suitably
+ written to handle Mac path names."""
+import os
+import panoptes_client
+from panoptes_client import SubjectSet, Panoptes
+from PIL import Image
+import requests
+
+while True:
+    User_name = input('Enter Zoonioverse User_name' + '\n')
+    Password = input('Entre Zooniverse Password - Warning it will show on the screen' + '\n')
+    try:
+        Panoptes.connect(username=User_name, password=Password)
+        break
+    except panoptes_client.panoptes.PanoptesAPIException:
+        print('Credentials not accepted')
+        retry = input('Enter "n" to cancel, any other key to retry' + '\n')
+        if retry.lower() == 'n':
+            quit()
+
+while True:
+    set_id = input('Entry subject set id to download:' + '\n')
+    try:
+        # check if the subject set exits
+        print('Please wait while I check this subject set exists and how many subjects are in it')
+        subject_set = SubjectSet.where(subject_set_id=set_id).next()
+        count_subjects = 0
+        subject_list = []
+        for subject in subject_set.subjects:
+            count_subjects += 1
+            subject_list.append(subject)
+            print(subject)
+        print('You have chosen to download ', len(subject_list), ' subjects from subject set ', set_id)
+        retry = input('Enter "n" to cancel this download, any other key to continue' + '\n')
+        if retry.lower() == 'n':
+            quit()
+        break
+    except panoptes_client.panoptes.PanoptesAPIException:
+        retry = input('Subject set not found, Enter "n" to cancel, any other key to retry' + '\n')
+        if retry.lower() == 'n':
+            quit()
+
+while True:
+    location = input('Enter the full path for the image directory to download to, or enter "." '
+                     'to use the current directory' + '\n')
+    if location == '.':
+        location = os.getcwd()
+        break
+    else:
+        if os.path.exists(location):
+            break
+        else:
+            print('That entry is not a valid path for an existing directory')
+            print('Example (Windows) "C:\Some_directory_name\Some_sub_directory"')
+            print('Note, no closing slash')
+            retry = input('Enter "n" to cancel, any other key to try again' + '\n')
+            if retry.lower() == 'n':
+                quit()
+              
+i =  0
+for item in subject_list:
+    try:
+        file_name = location + os.sep + item.metadata['Filename']
+    except KeyError:
+        file_name = location + os.sep + item.id + '.jpg'
+      
+    print(file_name)
+    if os.path.isfile(file_name):
+        print(file_name, ' already exists, not downloaded')
+        i += 1
+        continue
+
+    # acquire the image
+    try:
+        im = Image.open(requests.get(item.locations[0]['image/jpeg'], stream=True).raw)
```

Comment:
Are all images jpegs? the locations is an array of json objects with the mime type as the key. The currently list could be: `image/jpeg image/png image/gif`

more details from https://github.com/zooniverse/Panoptes/blob/cc6d7a5413e8097e0878cda6ce33af9ada128666/app/models/medium.rb#L13

Reply:
This is a serious one!  I am not sure how to handle it either - the exif extraction is not likely going to work with anything but jpg, jpeg and png, and I am not sure how to extract the mime type from the subject other than looking at .locations.
Since this script is likely only going to be used rarely to recover known file types maybe I should just ask for the file type to download and make it clear each run will only download files of the type specified...  There are other severe restrictions on the subjects this will work with - specifically they must be single-image, or at least the script only captures frame 0.

Reply:
we could have a list of allowed mime types (based on panoptes) that it looks for in the locations array, something like?
``` python
allowed_mimes = [ 'image/jpeg', 'image/png', 'image/gif' ]
for mime_type in allowed_mimes:
    for location in requests.get(item.locations):
        if mime_type in location:
            # note you'll probably have to deal with naming this resulting file with the location index suffix too.
            im = Image.open(location[mime_type], stream=True).raw)
```

Reply:
I'd suggest something like:

```python
im = Image.open(requests.get(item.locations[0].values(0), stream=True).raw)
```

Reply:
(Note that I've just edited that to fix a typo if you're reading this in email)